### DISCUSSION TITLE
422 vs 400

### DISCUSSION THREAD

**Discussion by fvanderelst** [fastapi/fastapi#2479] [Questions] [answered]
Hello,

Fantastic work on FastAPI.  Awesome.  What I've been waiting for for a long time !  I'll sponsor.

I've noticed that when validation fails, I get a 422 response.

I've spent lots of the last two years looking into 422 vs 400, and I believe that when a validation fails, 400 Bad request is a better code than 422.

Looking at https://httpstatuses.com/422, for instance:
"The server understands the content type of the request entity (hence a 415 Unsupported Media Type status code is inappropriate), and the syntax of the request entity is correct (thus a 400 Bad Request status code is inappropriate) but was unable to process the contained instructions."

whereas https://httpstatuses.com/400:
"The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing)."

I believe submitting a wrong-type parameter, say an "x" instead of a 5, means that the syntax of a request is incorrect.  It's a malformed request.  So it should yield a 400 rather than 422.

I've debated on this many times over the last few years at work, and ended up switching most cases of 422 to 400's.  I believe that's what most API writers out there tend to do.

I'll try to find "best practice" type articles and will forward them.

I realise it may not be easy/worthwhile changing this, as it's not really a "bug" - people's perceptions will vary.  What is your thought ?

Is it possible to configure what http code gets returned ?

I see https://fastapi.tiangolo.com/tutorial/handling-errors/#use-the-requestvalidationerror-body so I believe I will be able to change the behaviour.

Many thanks again, great work !

fve@vander-elst.org

**Comment by Kludex** [comment, marked as answer]
It's possible to configure, yes. The `exception handler` is your feature for that. But to change the values on the OpenAPI schema, you might need this: https://github.com/tiangolo/fastapi/issues/2455 .

Maybe we could improve the procedure and make it more customizable. :thinking:

**Comment by fvanderelst** [comment]
Thanks very much, I'll have a look at the exception handler, and the ticket.

Best regards

Freddie

**Comment by matez0** [comment]
Hi,
According to RFCs and [mozilla docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) and [discussions](https://stackoverflow.com/questions/16133923/400-vs-422-response-to-post-of-data) which also refer to RFCs,
the framework should respond **400** in case of payload schema validation error.

What was the design decision about responding 422 by default and why cannot **400** be used to be conform with the HTTP?
